src/main/java/htsjdk/samtools/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,674 @@
+        FastaReferenceWriterBuilder.checkBasesPerLine(basesPerLine);
+        closeSequence();
+        if (sequenceNamesAndSizes.containsKey(sequenceName)) {
+            throw new IllegalStateException("the input sequence name '" + sequenceName + "' has already been added");
+        }
+        currentSequenceName = sequenceName;
+        currentBasesPerLine = basesPerLine;
+        final StringBuilder builder = new StringBuilder(sequenceName.length() + nonNullDescription.length() + 2);
+        builder.append(HEADER_START_CHAR).append(sequenceName);
+        if (!nonNullDescription.isEmpty()) {
+            builder.append(HEADER_NAME_AND_DESCRIPTION_SEPARATOR).append(nonNullDescription);
+        }
+        fastaStream.write(builder.toString().getBytes(CHARSET));
+        fastaStream.write(LINE_SEPARATOR);
+        currentSequenceOffset = fastaStream.getBytesWritten();
+        return this;
+    }
+
+    private void closeSequence()
+            throws IOException {
+        if (currentSequenceName != null) {
+            if (currentBasesCount == 0) {
+                throw new IllegalStateException("no base was added");
+            }
+            sequenceNamesAndSizes.put(currentSequenceName, currentBasesCount);
+            writeIndexEntry();
+            writeDictEntry();
+            fastaStream.write(LINE_SEPARATOR);
+            currentBasesCount = 0;
+            currentLineBasesCount = 0;
+            currentSequenceName = null;
+        }
+    }
+
+    private void writeIndexEntry()
+            throws IOException {
+        indexWriter.append(currentSequenceName).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesCount)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentSequenceOffset)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesPerLine)).append(INDEX_FIELD_SEPARATOR_CHR)
+                .append(String.valueOf(currentBasesPerLine + LINE_SEPARATOR.length)).append(LINE_SEPARATOR_CHR);
+    }
+
+    private void writeDictEntry() {
+        dictCodec.encodeSequenceRecord(new SAMSequenceRecord(currentSequenceName, (int) currentBasesCount));
+    }
+
+    /**
+     * Adds bases to current sequence from a {@code byte} array.
+     *
+     * @param bases array containing the bases to be added.
+     * @return this instance.
+     * @throws IllegalArgumentException if {@bases} is {@code null} or
+     *                                  the input array contains invalid bases (as assessed by: {@link SequenceUtil#isIUPAC(byte)}).
+     * @throws IllegalStateException    if no sequence was started or the writer is already closed.
+     * @throws IOException              if such exception is throw when writing in any of the outputs.
+     */
+    public FastaReferenceWriter appendBases(final byte[] bases)
+            throws IOException {
+        return appendBases(bases, 0, bases.length);
+    }
+
+    /**
```

Comment:
I've stopped reviewing, because I just don't understand why this class doesn't leverage the various htsjdk classes for it's API.  For example, `ReferenceSequence` and `SAMSequenceRecord`.  I get wanting to feed sub-parts of the reference FASTA at a time, but can't that be handled by giving it an iterator over sub-sequences that's evaluated lazily?

Reply:
That sounds nice...but not a simple port. I was looking for something so that folks could write FASTA files with htsjdk.....however, I've realized that the length isn't used...so this is a set of names...I'll make that change.